### PR TITLE
Feat/storage provider admin UI

### PR DESCRIPTION
**Pull request by tnodell**
This implements Media Upload Support in Graphweaver, described in these tickets:

https://app.clickup.com/t/2704425/ID-4249
and
https://app.clickup.com/t/2704425/ID-6033

https://github.com/exogee-technology/graphweaver/assets/81122022/488f6215-9e95-47f7-adbc-6939d552c6e2

https://github.com/exogee-technology/graphweaver/assets/81122022/5ffc71eb-fbf2-41d3-9615-0e8415bfcb74

### CODE REVIEW

File: src/examples/s3-storage/src/backend/database.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import { PostgreSqlDriver } from '@mikro-orm/postgresql';
+
+import { Submission, User } from './entities';
+
```
Comment (tnodell) [inline src/examples/s3-storage/src/backend/database.ts:4]: I could remove all of the postgres and user things from this example.

File: src/examples/s3-storage/src/backend/schema/submission/entity.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+	@Field(() => ID)
+	id!: string;
+
+	@ReadOnlyProperty({ adminUI: true, backend: false })
+	@Field(() => String, { nullable: true })
+	key?: string;
+
+	// "key" must match the name of the field on the entity that gets the url from s3
+	@DownloadUrlField({ storageProvider: s3, key: 'key', mediaType: MediaTypes.IMAGE })
```
Comment (steve-keep) [inline src/examples/s3-storage/src/backend/schema/submission/entity.ts:32]: Is "key" the best name here I wonder if we could have a longer more explicit name?

File: src/packages/admin-ui-components/src/detail-panel/component.tsx [hunk 3]
```diff
@@ -253,6 +268,22 @@ export const DetailPanel = () => {
...
+					delete values.uploadUrl;
+					delete values.file;
+					// delete the value where the name is the field that has the image or media type
+					delete values[selectedEntity.fields.find((field) => field.type === 'Image')?.name ?? ''];
+					delete values[selectedEntity.fields.find((field) => field.type === 'Media')?.name ?? ''];
+				}
+
+				// if uploadUrl and downloadUrl are there, remove them because theyre not on SubmissionCreateOrUpdateInput
+				if (values.uploadUrl === null) delete values.uploadUrl;
+				if (values.downloadUrl === null || values.downloadUrl === '') delete values.downloadUrl;
```
Comment (steve-keep) [inline src/packages/admin-ui-components/src/detail-panel/component.tsx:284]: This wont always be `downloadUrl`, we need to make this generic.

File: src/packages/admin-ui-components/src/detail-panel/fields/media-field/media-field.tsx [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+		}
+	};
+
+	const handleOnDelete = () => {
+		setValues((prev: any) => ({
+			...prev,
+			key: null,
+			uploadUrl: null,
+			file: null,
+			downloadUrl: null,
```
Comment (steve-keep) [inline src/packages/admin-ui-components/src/detail-panel/fields/media-field/media-field.tsx:52]: Im not sure if `downloadUrl` refers to this https://github.com/exogee-technology/graphweaver/pull/274/files#diff-1c99eda95dafecacb997936b67bff1adacfe681874acc6e916aee0b64e03c7f3R33

Could you test changing the attribute name on the backend to confirm it works?

File: src/packages/admin-ui-components/src/detail-panel/graphql.ts [hunk 5]
```diff
@@ -39,3 +39,17 @@ export const getRelationshipQuery = (entityName: string, summaryField?: string)
...
+
+export const getUploadUrlMutation = gql`
+	mutation GetUploadUrl($key: ID!) {
+		getUploadUrl(key: $key)
+	}
+`;
+
+export const createSubmissionMutation = gql`
+	mutation Mutation($createSubmissionData: SubmissionInsertInput!) {
+		createSubmission(data: $createSubmissionData) {
```
Comment (steve-keep) [inline src/packages/admin-ui-components/src/detail-panel/graphql.ts:51]: Submission is example specific so this needs to be generic also.

File: src/packages/core/src/decorators/read-only-property.ts [hunk 6]
```diff
@@ -1,15 +1,31 @@
...
-		// as the target, so we need to as well for later.
-		Reflect.metadata(readOnlyPropertyKey, true)(target.constructor, propertyKey);
+		if (propertyKey) {
+			if (adminUI)
+				Reflect.metadata(readOnlyPropertyAdminUIKey, true)(target.constructor, propertyKey);
+			if (backend)
+				Reflect.metadata(readOnlyPropertyBackendKey, true)(target.constructor, propertyKey);
+			return;
+		}
+		if (adminUI) Reflect.metadata(readOnlyPropertyAdminUIKey, true)(target.constructor);
```
Comment (steve-keep) [inline src/packages/core/src/decorators/read-only-property.ts:20]: I think we should error if we get here as we do not want people decorating a class with this decorator?

File: src/packages/core/src/decorators/read-only.ts [hunk 7]
```diff
@@ -1,11 +1,28 @@
...
+type Props = {
+	backend?: boolean;
+	adminUI?: boolean;
+};
 
-export function isReadOnly(target: any) {
-	return !!Reflect.getMetadata(readOnlyKey, target);
+export function ReadOnly({ backend, adminUI }: Props = { backend: true, adminUI: true }) {
+	return (target: any, propertyKey?: string | symbol) => {
+		if (propertyKey) {
```
Comment (steve-keep) [inline src/packages/core/src/decorators/read-only.ts:13]: This block also needs to be removed for the same reason as above.

File: src/packages/graphweaver-storage-provider/package.json [hunk 8]
```diff
@@ -0,0 +1,39 @@
+{
+	"name": "@exogee/graphweaver-storage-provider",
+	"version": "0.2.11",
```
Comment (steve-keep) [inline src/packages/graphweaver-storage-provider/package.json:3]: Set to `0.2.14`

File: src/packages/graphweaver-storage-provider/src/decorators/download-url.ts [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+			complexity: undefined,
+		});
+
+		const fieldResolver = async (root: any) => {
+			// If the key is set to null, we don't want to return a download url, an return empty string
+			if (root[key] === null) return '';
+
+			// Check that key is a property on the object
+			if (!root[key]) {
+				console.log('root', root);
```
Comment (steve-keep) [inline src/packages/graphweaver-storage-provider/src/decorators/download-url.ts:91]: remove these or convert to logger

File: src/packages/graphweaver-storage-provider/src/storageProvider.ts [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+			Key: key,
+		});
+
+		const uploadURL = await getSignedUrl(s3, command, { expiresIn: this.expiresIn });
+		return uploadURL;
+	}
+}
+
+@Resolver()
+export class S3StorageProvider {
```
Comment (steve-keep) [inline src/packages/graphweaver-storage-provider/src/storageProvider.ts:56]: If we are going to have this as a resolver then we should call it one. Either merge with the above resolver code or remove the resolver.

**Review by steve-keep — COMMENTED** [review]
Looks great @tnodell, few comments below.

File: src/examples/s3-storage/src/backend/schema/index.ts [hunk 11]
```diff
@@ -0,0 +1,17 @@
+import { SubmissionResolver } from './submission';
+import { S3StorageResolver, StorageType } from '@exogee/graphweaver-storage-provider';
+import { Resolver } from '@exogee/graphweaver';
+
+@Resolver()
+class MyS3StorageResolver extends S3StorageResolver {
+	constructor() {
+		if (!process.env.AWS_S3_BUCKET) throw new Error('Missing required env AWS_S3_BUCKET');
+		super({
+			bucketName: process.env.AWS_S3_BUCKET,
```
[the remark this answers is not in the document]
Comment (tnodell) [reply src/examples/s3-storage/src/backend/schema/index.ts:10]: Yea you do, should we have this handle an array of buckets? How would one know which belongs to which entity?

[the remark this answers is not in the document]
Comment (tnodell) [reply src/examples/s3-storage/src/backend/schema/index.ts:10]: Will follow up with @thekevinbrown on this.

File: src/examples/s3-storage/src/backend/schema/submission/entity.ts (hunk 2 above)
Comment (tnodell) [reply src/examples/s3-storage/src/backend/schema/submission/entity.ts:32]: Will rename to 'resourceId'

File: src/packages/admin-ui-components/src/detail-panel/component.tsx (hunk 3 above)
Comment (tnodell) [reply src/packages/admin-ui-components/src/detail-panel/component.tsx:284]: Will rename the entity downloadUrl to something else to be clearer.

File: src/packages/admin-ui-components/src/detail-panel/fields/media-field/media-field.tsx (hunk 4 above)
Comment (tnodell) [reply src/packages/admin-ui-components/src/detail-panel/fields/media-field/media-field.tsx:52]: No these are diff, will change, see above.

File: src/packages/admin-ui-components/src/detail-panel/graphql.ts (hunk 5 above)
Comment (tnodell) [reply src/packages/admin-ui-components/src/detail-panel/graphql.ts:51]: Will follow pattern for getRelationshipQuery to dynamically generate.

File: src/packages/core/src/decorators/read-only-property.ts (hunk 6 above)
Comment (tnodell) [reply src/packages/core/src/decorators/read-only-property.ts:20]: Will delete

Comment (tnodell) [reply src/packages/core/src/decorators/read-only-property.ts:20]: will do

File: src/examples/s3-storage/src/backend/schema/index.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (thekevinbrown) [reply src/examples/s3-storage/src/backend/schema/index.ts:10]: Yup, current plan is you can create a single S3StorageResolver and import it from multiple entities, or you can create one per entity, e.g. whatever you need.

File: src/packages/admin-ui-components/src/detail-panel/graphql.ts (hunk 5 above)
Comment (tnodell) [reply src/packages/admin-ui-components/src/detail-panel/graphql.ts:51]: This is actually leftovers, this mutation is not needed, handled by createEntity mutation.

File: src/packages/end-to-end/src/__tests__/api/media/downloadUrl.test.ts [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+					}
+					__typename
+				}
+			}
+		`,
+	});
+	assert(response.body.kind === 'single');
+	const result = response.body.singleResult.data?.result as unknown as Schema;
+
+	console.log(JSON.stringify(result, null, 2));
```
Comment (steve-keep) [inline src/packages/end-to-end/src/__tests__/api/media/downloadUrl.test.ts:135]: Can this file be removed?

**Review by steve-keep — APPROVED** [review]

**Pull request merged**